crates/ibc/src/core/ics26_routing/context.rs

```diff
@@ -175,7 +175,7 @@ pub trait Module: Debug {
         Ok(ModuleExtras::empty())
     }
 
-    fn on_recv_packet_execute(
```

Comment:
I initially left it as `*_execute` (and without any `*_validate`) because it turned out there was [nothing to validate](https://github.com/cosmos/ibc-rs/blob/abff7a98d0bb3d36735829db3e8673b2f2f6a4d5/crates/ibc/src/core/context/recv_packet.rs#L31) in the token transfer app. However I'm now realizing that I also removed it from the `Module` trait, which is the interface for *all* apps.

The solution is to add back `on_recv_packet_validate()` (and keep this as `on_recv_packet_execute()`). Let's do it in a separate PR.

Reply:
#513

Reply:
I did so. Having separated val/exec, we end up in this [line](https://github.com/cosmos/ibc-rs/blob/116638e70636eefcc3d81ecf54c37df54b1a387a/crates/ibc/src/core/context.rs#L223) where we can't pass ctx as mutable, which forces us to touch some other parts. Give it another try?

Reply:
> This output is used for setting the ack state and emitting an event, which needs mutable context.

not it the validation step; this is done in the execution step. we want 

```rust
trait Module {
  // ...
  fn on_recv_packet_validate(
        &self,
        packet: &Packet,
        relayer: &Signer,
    ) -> Result<(), PacketError>; // Notice the return type; on ack nor events
  // ...
}
```